File: api/internal/reqid/reqid.go
// Package reqid carries the identifier that ties one request to every line it
// produced, in this service and in the next one.
//
// It is its own package, and forty lines is the right size for it. The
// middleware sets the id, the problem writer reads it, and the middleware
// already writes problems — so the two cannot both own it without a cycle.
// F1 adds a trace id beside it and touches neither.
package reqid

import (
	"context"
	"crypto/rand"
	"encoding/hex"
)

// Header is the name on the wire. ADR 0009: the same value appears as
// `requestId` in every problem document, so quoting either one from a bug
// report finds every log line from both services.
const Header = "X-Request-Id"

// Length in bytes before hex encoding. Sixteen gives 32 characters — long
// enough that a collision is not worth reasoning about, short enough to paste
// into a chat message and grep for.
const size = 16

type contextKey struct{}

// New returns a fresh identifier.
//
// crypto/rand.Read never fails on any platform Go supports; since Go 1.24 it
// panics rather than returning an error, so there is no degraded path here to
// get wrong quietly.
func New() string {
	var b [size]byte
	rand.Read(b[:]) //nolint:errcheck // documented never to fail; it panics instead
	return hex.EncodeToString(b[:])
}

// Valid reports whether an inbound identifier may be reused.
//
// This is not fussiness. The id is written into a JSON log line and echoed in a
// response header: a newline in it forges log entries, and a stray control
// character splits the header. Anything that does not match is replaced with a
// fresh one rather than rejected, because the caller's malformed header is not
// a reason to fail their request.
func Valid(s string) bool {
	if len(s) < 8 || len(s) > 64 {
		return false
	}
	for _, r := range s {
		switch {
		case r >= 'a' && r <= 'z',
			r >= 'A' && r <= 'Z',
			r >= '0' && r <= '9',
			r == '-', r == '_':
		default:
			return false
		}
	}
	return true
}

// With returns a context carrying the identifier.
func With(ctx context.Context, id string) context.Context {
	return context.WithValue(ctx, contextKey{}, id)
}

// From returns the identifier, or the empty string when there is none. An empty
// answer is a handler running outside the chain — a unit test, usually — and
// the callers treat it as "nothing to quote" rather than as an error.
func From(ctx context.Context) string {
	id, _ := ctx.Value(contextKey{}).(string)
	return id
}

File: web/lib/reqid.ts
// The identifier a visitor can quote, and the one every line of this container
// is filed under.
//
// Its own file rather than a corner of lib/trace.ts, and for the reason
// api/internal/reqid states about the same split: the two answer different
// questions. A request id is what somebody reads off an error page and pastes
// into a message; a trace id is what a backend joins on. They also have
// different futures — F6 replaces the trace generator with the OpenTelemetry SDK
// and leaves this untouched.
//
// THE ID IS ALWAYS MINTED HERE, NEVER ADOPTED.
//
// The api adopts an inbound X-Request-Id from a trusted proxy, because the web
// tier is meant to pass one through. Web has no such notion: it sits behind
// Traefik on the open internet, and every peer is a stranger. An adopted id
// would be a name a stranger picks for his own request in our log and in our
// answers to other people — the same argument middleware/requestid.go makes, and
// here it has no exception to carve out.
//
// `isValid` therefore has exactly one caller today: the test that proves the
// shape of what this mints. It exists because the shape is a promise — the value
// goes into a JSON log line and a response header, where a newline forges
// entries and a control character splits headers — and a promise with no check
// is a comment.

/**
 * The name on the wire. ADR 0009: the same value appears as `requestId` in every
 * problem document the api returns, so quoting either one finds the lines.
 */
export const REQUEST_ID_HEADER = "X-Request-Id";

// Sixteen bytes gives 32 characters — long enough that a collision is not worth
// reasoning about, short enough to paste into a chat message and grep for. The
// same size the api uses, so the two are indistinguishable in a log.
const SIZE = 16;

/** A fresh identifier: 32 lowercase hex characters. */
export function createRequestId(): string {
  const b = new Uint8Array(SIZE);
  crypto.getRandomValues(b);
  return Array.from(b, (x) => x.toString(16).padStart(2, "0")).join("");
}

/**
 * Whether an identifier may appear in a header and a log line.
 *
 * The same range the api accepts, so that a value one service tolerates is one
 * the other tolerates too: 8 to 64 characters of `[A-Za-z0-9_-]`.
 */
export function isValidRequestId(s: string): boolean {
  if (s.length < 8 || s.length > 64) return false;
  for (const c of s) {
    const ok =
      (c >= "a" && c <= "z") ||
      (c >= "A" && c <= "Z") ||
      (c >= "0" && c <= "9") ||
      c === "-" ||
      c === "_";
    if (!ok) return false;
  }
  return true;
}
